## *Jeu de données IRVE*
-------------------
# Analyse des relations entre champs (intégrité des données)
------------------------------
## Contexte
- clarification du rôle des modèles de données dans les jeux de données (cf mise à jour récente des [guides data.gouv](https://guides.etalab.gouv.fr/qualite/documenter-les-donnees/))
- intégration d'une propriété "relationship" dans les schémas de données ([issue TableSchema](https://github.com/frictionlessdata/specs/issues/803) en cours de validation)
- création d'outils de contrôle des relations entre champs des jeux de données tabulaires (cf usage ci-dessous)

## Objectifs
- valider sur un cas réel l'utilisation d'un modèle de données en complément d'un schéma de données
- identifier les apports que pourraient avoir les contrôles de validation des relations entre champs

## Résultats
- les principaux résultats sont regroupés sur un [tableau de bord](https://nbviewer.org/github/loco-philippe/Environmental-Sensing/blob/main/python/Validation/irve/Analyse/IRVE_indicateurs.ipynb) 
- la formalisation d'un modèle de données facilite la compréhension des données et des relations entre champs
- l'outil de contrôle permet d'améliorer significativement la qualité des données par l'identification d'incohérences de relations
- l'identification des incohérences permet de trouver des stratégies de réduction des écarts (dans l'exemple ci-dessous, on passe 36% d'écart à 3,8 %)  
- l'analyse des données permet de (re)construire le modèle de données qui minimise les incohérences
- les incohérences détectées sur le jeu de données IRVE restent faibles (inférieures à 5 % des point de charge documentés - voir chapitre 4)

## Suite à donner
- Mettre à jour, valider et publier le modèle de données IRVE
- Définir les contrôles supplémentaires à intégrer pour toutes nouvelles données ainsi que pour le jeu complet
- Mettre en oeuvre les outils de contrôle

## Evolutions possibles 
- Ajouter dans les guides d'Etalab un guide pour les modèles de données 
- Intégrer dans les schémas de données la propriété "relationship" en cours de validation,
- Définir un indicateur qui mesure l'écart (existant / attendu) des relations entre champs

## Sommaire
*(liens actifs sur jupyter Notebook ou Nbviewer)*
- [1 - modèle de données](#1---modèle-de-données)
- [2 - Initialisation](#2---Initialisation)
- [3 - Séparation des pdc itinerance et hors itinerance](#3---Séparation-des-pdc-itinerance-et-hors-itinerance)
- [4 - Bilan initial intégrité](#4---Bilan-initial-intégrité)
- [5 - Séparation doublons pdc - date de maj](#5---Séparation-doublons-pdc---date-de-maj)
- [6 - Séparation doublons station - date de maj](#6---Séparation-doublons-station---date-de-maj)
- [7 - Synthèse](#7---Synthèse)
- [8 - Exemples d'erreurs résiduelles](#8---Exemples-d\'erreurs-résiduelles)

Ce Notebook peut être consulté sur [nbviewer](http://nbviewer.org/github/loco-philippe/Environmental-Sensing/tree/main/python/Validation/irve/Analyse)

données utilisées : https://www.data.gouv.fr/fr/datasets/fichier-consolide-des-bornes-de-recharge-pour-vehicules-electriques/    
fichier : "*consolidation-etalab-schema-irve-statique-v-2.2.0-2023xxxx.csv*"

-----------------------
# 1 - modèle de données

Le modèle de données proposé ci-dessous est construit sur la base du schéma de données mis à disposition et du contenu du jeu de données.    
Il est à consolider en fonction de l'expertise des concepteurs et réutilisateurs (voir [guide méthodologique](https://github.com/loco-philippe/Environmental-Sensing/blob/main/property_relationship/FR_methodology.ipynb)).

*Notation:*
- *M : Mandatory - documentation obligatoire*
- *PK : Primary Key - identifiant unique de l'entité*
- *Root : champ fictif associé à une ligne du tableau*

In [1]:
from base64 import b64encode
from IPython.display import Image, display
with open('IRVE_modele.txt', 'r', encoding="utf-8") as f:
    modele = f.read()
display(Image(url="https://mermaid.ink/img/" + b64encode(modele.encode("ascii")).decode("ascii")))

In [2]:
def analyse_integrite(data, fields, pr=True):
    '''analyse les relations du DataFrame définies dans le dict fields et retourne un dict avec pour chaque contrôle la
    liste des index ko. Les résultats des contrôles sont également ajoutés sous forme de champs booléens à data'''
    analyse = Analysis(data)
    dic_res = analyse.check_relationship(fields)
    data['ok'] = True
    for name, lis in dic_res.items():
        data[name] = True
        data.loc[lis, name] = False
        data['ok'] = data['ok'] & data[name]
        if pr:
            print('{:<50} {:>5}'.format(name, len(data) - data[name].sum()))
    return dic_res

def add_nbre_pdc(data, name='ok', pr=True):
    '''ajoute un champ avec le nombre de pdc calculés et le compare au champ 'nbre_pdc' '''
    data['nb_pdc_calc'] = data.groupby('id_station_itinerance')['index'].transform('count')
    data['nbre_pdc_ok'] = data['nb_pdc_calc'] == data['nbre_pdc']    
    data['calc-nbre'] = data['nb_pdc_calc'] - data['nbre_pdc']
    #data[name] = data[name] & data['nbre_pdc_ok']
    if pr:
        ecart_nbre = len(data) - data['nbre_pdc_ok'].sum()
        print('{:<20} {:>5}'.format('nbre_pdc_ko', ecart_nbre))
        print(round(ecart_nbre / len(data) * 100), ' %')

------
# 2 - Initialisation
## initialisation des données
- lecture du fichier issu de l'api

In [3]:
from datetime import datetime
import json
from observation import Ilist, Analysis
import pandas as pd
dates = ['2023-03-03', '2023-04-17', '2023-05-02', '2023-05-24', '2023-06-17']
date= dates[-1]
log = {'date_irve': date, 
       'file': 'consolidation-etalab-schema-irve-statique-v-2.2.0-'+date[:4]+date[5:7]+date[8:]+'.csv',
      'chemin': 'https://raw.githubusercontent.com/loco-philippe/Environmental-Sensing/main/python/Validation/irve/Analyse/'}
#log = {'date_irve': '2022-06-06', 'file': 'consolidation-etalab-schema-irve-v-2.0.2-20220606-propre2.csv',
#      'chemin': 'D:\\philippe\\python ESstandard\\Environmental-Sensing\\python\\Validation\\irve\\'}
irve = pd.read_csv(log['chemin'] + log['file'], sep=',', low_memory=False)
log['len_irve'] = len(irve)
print('nombre de lignes : ', log['len_irve']) 

nombre de lignes :  84386


## schéma de données
Le schéma de données restreint à la propriété 'relationship' et construit à partir du modèle de données est le suivants :

In [4]:
# complément à inclure dans le schéma de données
fields = [
 
 # relation unicité des pdl
 { "name": "index",
   "relationship" : { "parent" : "id_pdc_itinerance", "link" : "coupled" }},   
 # relations inter entités
 { "name": "contact_operateur",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "nom_enseigne",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "coordonneesXY",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "id_station_itinerance",
   "relationship" : { "parent" : "id_pdc_itinerance",     "link" : "derived" }},
 # relations intra entité - station
 { "name": "nom_station",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "implantation_station",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 #{ "name": "date_maj",
 #  "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "nbre_pdc",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "condition_acces",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "horaires",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "station_deux_roues",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 # relations intra entité - localisation
 { "name": "adresse_station",
   "relationship" : { "parent" : "coordonneesXY",         "link" : "derived" }} ]

# liste des champs liés à un controle (relations) et obligatoires (mandatory)
relations = ['index', 'contact_operateur', 'nom_enseigne', 'coordonneesXY', 'adresse_station', 'id_station_itinerance', 
             'nom_station', 'implantation_station', 'nbre_pdc', 'condition_acces', 'horaires', 'station_deux_roues', 
             'id_pdc_itinerance', 'date_maj', 'last_modified']
mandatory =  ['contact_operateur', 'nom_enseigne', 'coordonneesXY', 'adresse_station', 'id_station_itinerance', 'nom_station',
           'implantation_station', 'nbre_pdc', 'condition_acces', 'horaires', 'station_deux_roues', 'id_pdc_itinerance', 
           'puissance_nominale', 'prise_type_ef', 'prise_type_2', 'prise_type_combo_ccs', 'prise_type_chademo', 
           'prise_type_autre', 'paiement_acte', 'paiement_autre', 'reservation',  'accessibilite_pmr', 'restriction_gabarit', 
           'date_maj', 'last_modified']

-----------------------
## 3 - Séparation des pdc itinerance et hors itinerance
- de l'ordre de 1 % des points de charge sont hors itinerance

In [5]:
# séparation des données
data = irve
data[['id_station_itinerance','id_pdc_itinerance']] = data[['id_station_itinerance','id_pdc_itinerance']].astype('string')
#data[['last_modified','date_maj']] = data[['last_modified','date_maj']].astype('datetime64')
data['non_concerne'] = data['id_station_itinerance'].str.contains('oncern') | data['id_pdc_itinerance'].str.contains('oncern')

non_concerne = data[data['non_concerne']].reset_index()['index']
itinerance = data[~data['non_concerne']].reset_index()
itinerance_init = itinerance.loc[:, relations]
log['pdc_hors_itinerance'] = len(non_concerne)
log['pdc_en_itinerance'] = len(itinerance)
print('nombre de pdc hors itinerance : ', log['pdc_hors_itinerance'])
print('nombre de pdc en itinerance   : ', log['pdc_en_itinerance'])

nombre de pdc hors itinerance :  815
nombre de pdc en itinerance   :  83571


In [6]:
# arborescence des champs et nombre de valeurs différentes
il = Ilist.obj(itinerance_init)
print(il.tree())

-1: root-derived (83571)
   0 : index (83571)
      1 : contact_operateur (134)
      2 : nom_enseigne (1271)
      3 : coordonneesXY (18591)
      4 : adresse_station (17740)
      5 : id_station_itinerance (20566)
      6 : nom_station (17509)
      7 : implantation_station (5)
      8 : nbre_pdc (52)
      9 : condition_acces (2)
      10: horaires (505)
      11: station_deux_roues (6)
      12: id_pdc_itinerance (49690)
      13: date_maj (630)
      14: last_modified (590)


-----------------------
## 4 - Bilan initial intégrité
- plus de 50 % des lignes présentent un défaut d'intégrité

In [7]:
# séparation données bonnes (itinerance_ok_1) et données résiduelles (itinerance_1)
res = analyse_integrite(itinerance_init, fields)
itinerance_ok_1 = itinerance_init.loc[itinerance_init.ok, relations].reset_index(drop=True)
itinerance_1 = itinerance_init.loc[~itinerance_init.ok, relations].reset_index(drop=True)
itinerance_init = itinerance_init.loc[:, relations]
log['init_ok'] = len(itinerance_ok_1)
log['init_ko'] = len(itinerance_1)
print("\nnombre d'enregistrements sans erreurs : ", log['init_ok'])
print("nombre d'enregistrements avec au moins une erreur : ", log['init_ko'])
print("taux d'erreur : ", round(log['init_ko'] / log['pdc_en_itinerance'] * 100), ' %')

index - id_pdc_itinerance                          42543
contact_operateur - id_station_itinerance          13317
nom_enseigne - id_station_itinerance                9856
coordonneesXY - id_station_itinerance              15099
id_station_itinerance - id_pdc_itinerance           1278
nom_station - id_station_itinerance                 3953
implantation_station - id_station_itinerance        1279
nbre_pdc - id_station_itinerance                    5107
condition_acces - id_station_itinerance               31
horaires - id_station_itinerance                   12140
station_deux_roues - id_station_itinerance         14103
adresse_station - coordonneesXY                     2067

nombre d'enregistrements sans erreurs :  36590
nombre d'enregistrements avec au moins une erreur :  46981
taux d'erreur :  56  %


-----------------------
## 5 - Séparation doublons pdc - date de maj
- plus de la moitié des pdc en erreur sont liées aux doublons de pdc
- la suppression des doublons permet de réduire de plus de 80% le nombre de lignes erronnées

In [8]:
# séparation doublons pdc (doublons_pdc) et données résiduelles (itinerance_2)
itinerance_1['doublons_pdc'] = itinerance_1.sort_values(by=['date_maj', 'last_modified']).duplicated('id_pdc_itinerance', keep='last')

doublons_pdc = itinerance_1[itinerance_1['doublons_pdc']].loc[:, relations].reset_index(drop=True)['index']
itinerance_2 = itinerance_1[~itinerance_1['doublons_pdc']].loc[:, relations].reset_index(drop=True)
itinerance_1 = itinerance_1.loc[:, relations]
log['doublons_pdc'] = len(doublons_pdc)
log['sans_doublons_pdc'] = len(itinerance_2)
print('nombre de doublons pdc : ', log['doublons_pdc'], ' soit ', round(log['doublons_pdc']/log['init_ko'] * 100), ' %')
print('nombre de pdc sans doublon   : ', log['sans_doublons_pdc'])

nombre de doublons pdc :  33881  soit  72  %
nombre de pdc sans doublon   :  13100


In [9]:
# séparation données bonnes (itinerance_ok_3) et données résiduelles (itinerance_3)
res = analyse_integrite(itinerance_2, fields)
itinerance_ok_3 = itinerance_2.loc[itinerance_2.ok, relations].reset_index(drop=True)
itinerance_3 = itinerance_2.loc[~itinerance_2.ok, relations].reset_index(drop=True)
itinerance_2 = itinerance_2.loc[:, relations]
log['etape3_ok'] = len(itinerance_ok_3)
log['etape3_ko'] = len(itinerance_3)
print("\nnombre d'enregistrements sans erreurs : ", log['etape3_ok'])
print("nombre d'enregistrements avec au moins une erreur : ", log['etape3_ko'])

index - id_pdc_itinerance                              0
contact_operateur - id_station_itinerance            335
nom_enseigne - id_station_itinerance                 507
coordonneesXY - id_station_itinerance               1504
id_station_itinerance - id_pdc_itinerance              0
nom_station - id_station_itinerance                 1682
implantation_station - id_station_itinerance         202
nbre_pdc - id_station_itinerance                    1194
condition_acces - id_station_itinerance               13
horaires - id_station_itinerance                     267
station_deux_roues - id_station_itinerance           770
adresse_station - coordonneesXY                      997

nombre d'enregistrements sans erreurs :  8756
nombre d'enregistrements avec au moins une erreur :  4344


-----------------------
## 6 - Séparation doublons station - date de maj
- plus de 15% des erreurs résiduelles sont liées au mélange d'anciens et de nouveaux pdc
- la suppression des anciens pdc permet de réduire de plus de 20% le nombre de lignes erronnées
- les dernières erreurs correspondent à des stations associées à 33 opérateurs et sont liées à des causes multiples

In [10]:
# séparation doublons stations (doublons_stat_maj) et données résiduelles (itinerance_4)
itinerance_3['stat_maj'] = itinerance_3.id_station_itinerance + itinerance_3.date_maj
stat_maj_unique = itinerance_3.sort_values(by='stat_maj').drop_duplicates('id_station_itinerance', keep='last')
itinerance_3['last_stat_maj'] = itinerance_3['stat_maj'].isin(stat_maj_unique['stat_maj'])

doublons_stat_maj = itinerance_3[~itinerance_3['last_stat_maj']].loc[:, relations].reset_index(drop=True)['index']
itinerance_4 = itinerance_3[itinerance_3['last_stat_maj']].loc[:, relations].reset_index(drop=True)
itinerance_3 = itinerance_3.loc[:, relations]
log['doublons_station'] = len(doublons_stat_maj)
log['sans_doublons_station'] = len(itinerance_4)
print('nombre de doublons stations : ', log['doublons_station'], ' soit ', 
      round(log['doublons_station']/log['etape3_ko'] * 100), ' %')
print('nombre de pdc sans doublon   : ', log['sans_doublons_station'])

nombre de doublons stations :  468  soit  11  %
nombre de pdc sans doublon   :  3876


In [11]:
# séparation données bonnes (itinerance_ok_5) et données résiduelles (itinerance_5 / itinerance_5_full)
res = analyse_integrite(itinerance_4, fields)
itinerance_ok_5 = itinerance_4.loc[itinerance_4.ok, relations].reset_index(drop=True)
itinerance_5_full = itinerance_4.loc[~itinerance_4.ok].reset_index(drop=True)
itinerance_5 = itinerance_5_full.loc[:, relations]
#itinerance_4 = itinerance_4.loc[:, relations]
log['etape5_ok'] = len(itinerance_ok_5)
log['etape5_ko'] = len(itinerance_5)
print("\nnombre d'enregistrements sans erreurs : ", log['etape5_ok'])
print("nombre d'enregistrements avec au moins une erreur : ", log['etape5_ko'], ' soit ', 
      round(log['etape5_ko']/log['etape3_ko'] * 100), ' %')

index - id_pdc_itinerance                              0
contact_operateur - id_station_itinerance             13
nom_enseigne - id_station_itinerance                 263
coordonneesXY - id_station_itinerance                909
id_station_itinerance - id_pdc_itinerance              0
nom_station - id_station_itinerance                 1396
implantation_station - id_station_itinerance          22
nbre_pdc - id_station_itinerance                    1076
condition_acces - id_station_itinerance                8
horaires - id_station_itinerance                      52
station_deux_roues - id_station_itinerance           448
adresse_station - coordonneesXY                      830

nombre d'enregistrements sans erreurs :  227
nombre d'enregistrements avec au moins une erreur :  3649  soit  84  %


In [12]:
# structure des données résiduelles et nombre de valeurs
il = Ilist.obj(itinerance_5)
print(il.tree())

-1: root-derived (3649)
   0 : index (3649)
      1 : contact_operateur (38)
      2 : nom_enseigne (337)
      3 : coordonneesXY (1174)
      4 : adresse_station (1179)
      5 : id_station_itinerance (1148)
         13: date_maj (88)
      6 : nom_station (1731)
      7 : implantation_station (5)
      8 : nbre_pdc (22)
      9 : condition_acces (2)
      10: horaires (70)
      11: station_deux_roues (6)
      12: id_pdc_itinerance (3649)
      14: last_modified (68)


In [13]:
# nb maxi d'erreurs
itinerance_4['somme'] = 0
for name in res.keys():
    itinerance_4['somme'] += 1 - itinerance_4[name]
erreurs = max(itinerance_4['somme'])
maxi = itinerance_4[itinerance_4.somme >= erreurs]
print("nombre d'enregistrements avec ", erreurs, " erreurs : ", len(maxi))

nombre d'enregistrements avec  4  erreurs :  66


-----------------------
## 7 - Synthèse


### fichiers
Génération des fichiers intégrant les défauts d'intégrité :
- fichier csv des lignes résiduelles à traiter (IRVE_itinerance_residuel)
- fichier csv des données itinerance avec indicateur des données à corriger ou à ignorer (IRVE_itinerance_complet)
- fichier csv des données itinerance valides (IRVE_itinerance_valide)
- fichier csv des doublons (IRVE_itinerance_doublons)

In [14]:
# consolidation des données
itinerance['doublons_stat_maj'] = itinerance['index'].isin(doublons_stat_maj)
itinerance['doublons_pdc'] = itinerance['index'].isin(doublons_pdc)
itinerance['lignes_a_corriger'] = itinerance['index'].isin(itinerance_5['index'])
itinerance['doublons_a_supprimer'] = itinerance['doublons_stat_maj'] | itinerance['doublons_pdc']
itinerance['lignes_ko'] = itinerance['doublons_a_supprimer'] | itinerance['lignes_a_corriger']
print('total des lignes à corriger : ', itinerance['lignes_a_corriger'].sum())
itinerance_doublons = itinerance[itinerance['doublons_a_supprimer']].reset_index(drop=True)
print('total des doublons à supprimer : ', len(itinerance_doublons))
itinerance_ok = itinerance[~itinerance['lignes_ko']].reset_index(drop=True)
print('nombre de pdc avec controles ok : ', len(itinerance_ok))

total des lignes à corriger :  3649
total des doublons à supprimer :  34349
nombre de pdc avec controles ok :  45573


In [15]:
#génération des fichiers
extension = log['date_irve'] +'.csv'
itinerance_5_full.to_csv('IRVE_itinerance_residuel' + extension)
itinerance.to_csv('IRVE_itinerance_complet' + extension)
itinerance_ok.to_csv('IRVE_itinerance_valide' + extension)
itinerance_doublons.to_csv('IRVE_itinerance_doublons' + extension)
log['IRVE_itinerance_residuel' + extension] = len(itinerance_5_full)
log['IRVE_itinerance_complet' + extension] = len(itinerance)
log['IRVE_itinerance_valide' + extension] = len(itinerance_ok)
log['IRVE_itinerance_valide_stat' + extension] = len(itinerance_ok.drop_duplicates('id_station_itinerance', keep='last'))
log['IRVE_itinerance_doublons' + extension] = len(itinerance_doublons)

### vérification de l'intégrité

In [16]:
# vérification de l'absence d'erreurs
res = analyse_integrite(itinerance_ok.loc[:, relations], fields, pr=False)
log['bilan_erreurs'] = sum([len(controle) for controle in res.values()])
log['date'] = datetime.now().isoformat()
print('bilan intégrité :')
print('    erreurs : ', log['bilan_erreurs'])

bilan intégrité :
    erreurs :  0


In [17]:
# structure des données bonnes
il = Ilist.obj(itinerance_ok.loc[:, mandatory])
print(il.tree())

-1: root-derived (45573)
   11: id_pdc_itinerance (45573)
      4 : id_station_itinerance (19184)
         1 : nom_enseigne (941)
         2 : coordonneesXY (14976)
            0 : contact_operateur (116)
            3 : adresse_station (14332)
            8 : condition_acces (2)
         5 : nom_station (15046)
         6 : implantation_station (5)
         7 : nbre_pdc (52)
         9 : horaires (440)
         10: station_deux_roues (5)
         19: paiement_autre (7)
      12: puissance_nominale (84)
      13: prise_type_ef (8)
      14: prise_type_2 (8)
      15: prise_type_combo_ccs (7)
      16: prise_type_chademo (7)
      17: prise_type_autre (7)
      18: paiement_acte (8)
      20: reservation (6)
      21: accessibilite_pmr (4)
      22: restriction_gabarit (104)
      23: date_maj (468)
      24: last_modified (395)


### Indicateurs

In [18]:
# génération des indicateurs liés aux règles
regles = ['Pdc non unique', 'Station multi-operateurs', 'Station multi-enseignes', 'Station multi-localisations', 
          'Pdc multi-stations', 'station avec plusieurs noms', 'station multi-implantations', 
          'nombre de pdc par station incoherent', 'station multi-acces', 'station multi-horaires', 
          'acces deux-roues incoherent', 'localisation multi-adresses']
principal = [16, 17, 18, 19, 20]
secondaire = [21, 22, 23, 24, 25, 26, 27]
irve = itinerance_5_full # residuel
total = len(irve)
indic = {}

irve['principal'] = True
for ind in principal:
    irve['principal'] &= irve.iloc[:,ind]
    indic[regles[ind-16]] = int(total - irve.iloc[:,ind].sum())
irve['secondaire'] = True
for ind in secondaire:
    irve['secondaire'] &= irve.iloc[:,ind]
    indic[regles[ind-16]] = int(total - irve.iloc[:,ind].sum())
irve['secondaire'] |= (~irve['principal'] & ~irve['secondaire'])
irve['verif'] = irve['principal'] & irve['secondaire']
indic['principal pdc'] = int(total - irve['principal'].sum())
indic['secondaire pdc'] = int(total - irve['secondaire'].sum())

irve_p = irve[~irve['principal']].drop_duplicates('id_station_itinerance').reset_index(drop=True)
irve_s = irve[~irve['secondaire']].drop_duplicates('id_station_itinerance').reset_index(drop=True)
indic['principal stat'] = len(irve_p)
indic['secondaire stat'] = len(irve_s)



In [19]:
# ajout des principaux opérateurs contributeur de défauts
operateurs = list(itinerance_5.drop_duplicates('contact_operateur')['contact_operateur'])
erreurs_op = [len(itinerance_5.loc[itinerance_5.contact_operateur == operateur]) for operateur in operateurs]
err_op, oper = tuple(zip(*(sorted(zip(erreurs_op, operateurs), reverse=True))[:5]))
indic['operateurs'] = list(oper)
indic['erreurs_operateurs'] = list(err_op)

print(len(err_op), ' opérateurs représentent : ', round(sum(err_op) / sum(erreurs_op) * 100), ' % des écarts\n')
for err, op in zip(indic['erreurs_operateurs'], indic['operateurs']):
    print(op, err)

5  opérateurs représentent :  73  % des écarts

support@virta.global 1026
hello@powerdot.fr 618
supervision-ev.france@totalenergies.com 506
pascal.lhermitte@siege27.fr 258
web@freshmile.com 257


In [20]:
# ajout de l'indicateur d'écart entre nombre de pdc calculés et saisis
add_nbre_pdc(irve, pr=True)
indic['nb_pdc_calc-nbre_pdc'] = int(irve['calc-nbre'].sum())

nbre_pdc_ko           1500
41  %


In [21]:
# stockage des indicateurs
log |= indic
with open('logfile.txt', 'a', encoding="utf-8") as f:
    f.write(json.dumps(log) + '\n')

-----------------------
## 8 - Exemples d'erreurs résiduelles 

### Erreurs multiples 
- exemple avec le nombre maximal d'erreurs (4) (2 stations soit 29 pdc avec : adresse, nom, nbre_pdc et deux-roues erronés)

In [22]:
maxi.loc[:, relations]

index     contact_operateur          nom_enseigne  \
3191  80164  support@virta.global                MODULO   
3192  80165  support@virta.global                MODULO   
3193  80166  support@virta.global                MODULO   
3194  80167  support@virta.global                MODULO   
3195  80168  support@virta.global                MODULO   
...     ...                   ...                   ...   
3325  80309  support@virta.global                MODULO   
3326  80310  support@virta.global                MODULO   
3651  84035      contact@e55c.com  ELECTRIC 55 CHARGING   
3652  84036      contact@e55c.com  ELECTRIC 55 CHARGING   
3792  84236      contact@e55c.com  ELECTRIC 55 CHARGING   

                coordonneesXY                adresse_station  \
3191  [4.57827600,49.4743290]              Place de la Halle   
3192  [4.38327800,49.3957250]              Place de la Halle   
3193  [4.38327800,49.3957250]              50 Route d'Hirson   
3194  [4.41915300,49.6982130]              50 Route d'Hirson   
3195  [4.41915300,49.6982130]                Rue de Raucourt   
...                       ...                            ...   
3325  [4.49431027,49.7880936]                 Gare de Lucquy   
3326  [4.70237100,49.9925440]                 Gare de Lucquy   
3651    [2.130358, 48.801996]     AVENUE DE PARIS VERSAILLES   
3652    [2.130358, 48.801996]     AVENUE DE PARIS VERSAILLES   
3792    [2.130358, 48.801996]  6 AVENUE DE PARIS -VERSAILLES   

     id_station_itinerance                                    nom_station  \
3191           FRS08P08354  RAUCOURT ET FLABA - 57540 - Place de la Halle   
3192           FRS08P08354  RAUCOURT ET FLABA - 57539 - Place de la Halle   
3193           FRS08P08282       MAUBERT FONTAINE -57530 - Route d'Hirson   
3194           FRS08P08282      MAUBERT FONTAINE - 57529 - Route d'Hirson   
3195           FRS08P08357   REMILLY AILLICOURT - 57520 - Rue de Raucourt   
...                    ...                                            ...   
3325           FRS08P08262                          LUCQUY - 57626 - Gare   
3326           FRS08P08262                          LUCQUY - 57625 - Gare   
3651           FR55CP78000                     AVENUE DE PARIS VERSAILLES   
3652           FR55CP78000                     AVENUE DE PARIS VERSAILLES   
3792           FR55CP78000                  6 AVENUE DE PARIS -VERSAILLES   

     implantation_station  nbre_pdc condition_acces horaires  \
3191               Voirie         1     Accès libre     24/7   
3192               Voirie         2     Accès libre     24/7   
3193               Voirie         1     Accès libre     24/7   
3194               Voirie         2     Accès libre     24/7   
3195               Voirie         1     Accès libre     24/7   
...                   ...       ...             ...      ...   
3325               Voirie         1     Accès libre     24/7   
3326               Voirie         2     Accès libre     24/7   
3651               Voirie         4   Accès réservé     24/7   
3652               Voirie         4   Accès réservé     24/7   
3792               Voirie         2   Accès réservé     24/7   

     station_deux_roues            id_pdc_itinerance    date_maj  \
3191              False                  FRS08E57540  2023-05-31   
3192              False                  FRS08E57539  2023-05-31   
3193              False                  FRS08E57530  2023-05-31   
3194              False                  FRS08E57529  2023-05-31   
3195              False                  FRS08E57520  2023-05-31   
...                 ...                          ...         ...   
3325              False                  FRS08E57626  2023-05-31   
3326              False                  FRS08E57625  2023-05-31   
3651              FALSE  FR55CE780004880188021299331  2022-11-03   
3652              FALSE  FR55CE780004880188021299351  2022-11-03   
3792              FALSE      FR55CEFR78000AVDEPAR1S1  2022-11-03   

                 

### Cohérence implantation_station - id_station
- 64 pdc sont liés à une erreur de choix d'implantation (3 stations, 1 opérateur)

In [23]:
itinerance_4.loc[~itinerance_4['implantation_station - id_station_itinerance'], relations]

index     contact_operateur          nom_enseigne  \
1693  64108  contact@qovoltis.com              QOVOLTIS   
1696  64219  contact@qovoltis.com              QOVOLTIS   
1697  64314  contact@qovoltis.com              QOVOLTIS   
1698  64345  contact@qovoltis.com              QOVOLTIS   
1699  64392  contact@qovoltis.com              QOVOLTIS   
1700  64394  contact@qovoltis.com              QOVOLTIS   
1704  64496  contact@qovoltis.com              QOVOLTIS   
1705  64506  contact@qovoltis.com              QOVOLTIS   
1706  64507  contact@qovoltis.com              QOVOLTIS   
1707  64508  contact@qovoltis.com              QOVOLTIS   
3623  83994      contact@e55c.com  ELECTRIC 55 CHARGING   
3625  83996      contact@e55c.com  ELECTRIC 55 CHARGING   
3626  83997      contact@e55c.com  ELECTRIC 55 CHARGING   
3631  84006      contact@e55c.com  ELECTRIC 55 CHARGING   
3730  84131      contact@e55c.com  ELECTRIC 55 CHARGING   
3731  84132      contact@e55c.com  ELECTRIC 55 CHARGING   
3791  84234      contact@e55c.com  ELECTRIC 55 CHARGING   
3794  84238      contact@e55c.com  ELECTRIC 55 CHARGING   
3835  84316      contact@e55c.com  ELECTRIC 55 CHARGING   
3836  84317      contact@e55c.com  ELECTRIC 55 CHARGING   
3837  84318      contact@e55c.com  ELECTRIC 55 CHARGING   
3850  84337      contact@e55c.com  ELECTRIC 55 CHARGING   

              coordonneesXY  \
1693   [-2.76491, 48.49415]   
1696   [-2.76491, 48.49415]   
1697   [-2.76491, 48.49415]   
1698   [-2.76491, 48.49415]   
1699   [-2.76491, 48.49415]   
1700   [-2.76491, 48.49415]   
1704   [-2.76491, 48.49415]   
1705   [-2.76491, 48.49415]   
1706   [-2.76491, 48.49415]   
1707   [-2.76491, 48.49415]   
3623  [6.640612, 43.269507]   
3625   [6.635375, 43.27044]   
3626   [6.635375, 43.27044]   
3631  [6.640612, 43.269507]   
3730   [6.635375, 43.27044]   
3731   [6.635375, 43.27044]   
3791  [6.636581, 43.272524]   
3794  [6.643206, 43.268741]   
3835  [7.130067, 43.641297]   
3836  [7.130067, 43.641297]   
3837  [7.130067, 43.641297]   
3850  [7.130067, 43.641297]   

                                        adresse_station id_station_itinerance  \
1693  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1696  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1697  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1698  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1699  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1700  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1704  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1705  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1706  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1707  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
3623            PARKING ESPACE DES LICES - SAINT TROPEZ           FR55CP83990   
3625                        PARKING PORT - SAINT TROPEZ           FR55CP83990   
3626                        PARKING PORT - SAINT TROPEZ           FR55CP83990   
3631            PARKING ESPACE DES LICES - SAINT TROPEZ           FR55CP83990   
3730                        PARKING PORT - SAINT TROPEZ           FR55CP83990   
3731                        PARKING PORT - SAINT TROPEZ           FR55CP83990   
3791                        CAPITAINERIE - SAINT TROPEZ           FR55CP83990   
3794             19 RUE DE LA RESISTANCE - SAINT TROPEZ           FR55CP83990   
3835                      COOLWORK - VILLENEUVE-LOUBET            FR55C06270V   
3836                      COOLWORK - VILLENEUVE-LOUBET            FR55C06270V   
3837                      COOLWORK - VILLENEUVE-LOUBET            FR55C06270V   
3850                      COOLWORK - VILLENEUVE-LOUBET            FR55C06270V   

                                        nom_station  \
1693  52 RUE DU

### Cohérence nom_station - id_station
- 328 pdc sont associés à une station avec un nom non cohérent

In [24]:
itinerance_4.loc[~itinerance_4['nom_station - id_station_itinerance'], relations]

index                        contact_operateur          nom_enseigne  \
67     8116  supervision-ev.france@totalenergies.com          INTERPARKING   
68     8117  supervision-ev.france@totalenergies.com          INTERPARKING   
69     8118  supervision-ev.france@totalenergies.com          INTERPARKING   
70     8119  supervision-ev.france@totalenergies.com          INTERPARKING   
71     8120  supervision-ev.france@totalenergies.com          INTERPARKING   
...     ...                                      ...                   ...   
3871  84378                         contact@e55c.com  ELECTRIC 55 CHARGING   
3872  84379                         contact@e55c.com  ELECTRIC 55 CHARGING   
3873  84380                         contact@e55c.com  ELECTRIC 55 CHARGING   
3874  84381                         contact@e55c.com  ELECTRIC 55 CHARGING   
3875  84385                         contact@e55c.com  ELECTRIC 55 CHARGING   

              coordonneesXY                          adresse_station  \
67    [7.265881, 43.695113]        1 Promenade des anglais 6000 NICE   
68    [7.265881, 43.695113]        1 Promenade des anglais 6000 NICE   
69    [7.265881, 43.695113]        1 Promenade des anglais 6000 NICE   
70    [7.265881, 43.695113]        1 Promenade des anglais 6000 NICE   
71    [7.265881, 43.695113]        1 Promenade des anglais 6000 NICE   
...                     ...                                      ...   
3871  [2.267834, 48.808019]             AVENUE JEAN JAURES - CLAMART   
3872  [2.267834, 48.808019]             AVENUE JEAN JAURES - CLAMART   
3873  [2.267834, 48.808019]             AVENUE JEAN JAURES - CLAMART   
3874  [2.267834, 48.808019]             AVENUE JEAN JAURES - CLAMART   
3875  [6.534893, 43.170946]  PARKING MEDIATHEQUE - CAVALAIRE SUR MER   

     id_station_itinerance                              nom_station  \
67               FRIPKNRME       INTERPARKING  - Nice Ruhl Méridien   
68               FRIPKNRME       INTERPARKING  - Nice Ruhl Méridien   
69               FRIPKNRME       INTERPARKING  - Nice Ruhl Méridien   
70               FRIPKNRME       INTERPARKING  - Nice Ruhl Méridien   
71               FRIPKNRME       INTERPARKING  - Nice Ruhl Méridien   
...                    ...                                      ...   
3871           FR55CP92140             AVENUE JEAN JAURES - CLAMART   
3872           FR55CP92140             AVENUE JEAN JAURES - CLAMART   
3873           FR55CP92140             AVENUE JEAN JAURES - CLAMART   
3874           FR55CP92140             AVENUE JEAN JAURES - CLAMART   
3875           FR55CP83240  PARKING MEDIATHEQUE - CAVALAIRE SUR MER   

              implantation_station  nbre_pdc condition_acces horaires  \
67    Parking privé à usage public        22     Accès libre     24/7   
68    Parking privé à usage public        22     Accès libre     24/7   
69    Parking privé à usage public        22     Accès libre     24/7   
70    Parking privé à usage public        22     Accès libre     24/7   
71    Parking privé à usage public        22     Accès libre     24/7   
...                            ...       ...             ...      ...   
3871                        Voirie         5   Accès réservé     24/7   
3872                        Voirie         5   Accès réservé     24/7   
3873                        Voirie         5   Accès réservé     24/7   
3874                        Voirie         5   Accès réservé     24/7   
3875                        Voirie         2   Accès réservé     24/7   

     station_deux_roues        id_pdc_itinerance    date_maj  \
67                false             FRIPKENRM012  2023-06-02   
68                false             FRIPKENRM011  2023-06-02   
69                false             FRIPKENRM010  2023-06-02   
70                false             FRIPKENRM009  2023-06-02   
71                false             FRIPKENRM008  2023-06-02   
...                 ...                      ...         ...   
3871              FALSE  F

### Cohérence adresse - coordonnées
- 440 pdc ont une adresse non cohérente avec les coordonnées géographiques

In [25]:
itinerance_4.loc[~itinerance_4['adresse_station - coordonneesXY'], relations]

index contact_operateur          nom_enseigne          coordonneesXY  \
0       826    sav@izivia.com                SIGEIF     [2.27736, 48.9917]   
1       827    sav@izivia.com                SIGEIF     [2.27736, 48.9917]   
2       830    sav@izivia.com                SIGEIF     [2.27736, 48.9917]   
3       831    sav@izivia.com                SIGEIF     [2.27736, 48.9917]   
4      2331    sav@izivia.com              SOREGIES    [0.30007, 46.54829]   
...     ...               ...                   ...                    ...   
3797  84241  contact@e55c.com  ELECTRIC 55 CHARGING  [5.976272, 45.526596]   
3798  84242  contact@e55c.com  ELECTRIC 55 CHARGING  [5.976272, 45.526596]   
3799  84243  contact@e55c.com  ELECTRIC 55 CHARGING  [5.976272, 45.526596]   
3869  84376  contact@e55c.com  ELECTRIC 55 CHARGING  [5.976272, 45.526596]   
3870  84377  contact@e55c.com  ELECTRIC 55 CHARGING  [5.976272, 45.526596]   

                                        adresse_station id_station_itinerance  \
0               RUE CRISTINO GARCIA PLACE ETIENNE DOLET       FRSIGPSIGE34211   
1               RUE CRISTINO GARCIA PLACE ETIENNE DOLET       FRSIGPSIGE34212   
2                                   RUE CRISTINO GARCIA       FRSIGPSIGE34412   
3                                   RUE CRISTINO GARCIA       FRSIGPSIGE34411   
4                                     RUE DE LÉCORCERIE       FRS86PSOREB1511   
...                                                 ...                   ...   
3797           BP - AIRE DE L'ABIS SAINT-JEOIRE-PRIEURE         FR55CPBP73190   
3798           BP - AIRE DE L'ABIS SAINT-JEOIRE-PRIEURE         FR55CPBP73190   
3799           BP - AIRE DE L'ABIS SAINT-JEOIRE-PRIEURE         FR55CPBP73190   
3869  BP - AIRE DE L'ABIS - SAINT-JEOIRE-PRIEURE - 7...           FR55CPBP731   
3870  BP - AIRE DE L'ABIS - SAINT-JEOIRE-PRIEURE - 7...           FR55CPBP731   

                                            nom_station  \
0     SIGEIF - 4 RUE CRISTINO GARCIA PLACE ETIENNE D...   
1     SIGEIF - 4 RUE CRISTINO GARCIA PLACE ETIENNE D...   
2     SIGEIF - 4 RUE CRISTINO GARCIA PLACE ETIENNE D...   
3     SIGEIF - 4 RUE CRISTINO GARCIA PLACE ETIENNE D...   
4                                      POITIERS SUD HPC   
...                                                 ...   
3797   BP - AIRE DE L'ABIS SAINT-JEOIRE-PRIEURE - 73190   
3798   BP - AIRE DE L'ABIS SAINT-JEOIRE-PRIEURE - 73190   
3799   BP - AIRE DE L'ABIS SAINT-JEOIRE-PRIEURE - 73190   
3869  BP - AIRE DE L'ABIS - SAINT-JEOIRE-PRIEURE - 7...   
3870  BP - AIRE DE L'ABIS - SAINT-JEOIRE-PRIEURE - 7...   

                     implantation_station  nbre_pdc condition_acces horaires  \
0                                  Voirie         2     Accès libre     24/7   
1                                  Voirie         2     Accès libre     24/7   
2                                  Voirie         2     Accès libre     24/7   
3                                  Voirie         2     Accès libre     24/7   
4                                  Voirie         1     Accès libre     24/7   
...                                   ...       ...             ...      ...   
3797  Station dédiée à la recharge rapide         8   Accès réservé     24/7   
3798  Station dédiée à la recharge rapide         8   Accès réservé     24/7   
3799  Station dédiée à la recharge rapide         8   Accès réservé     24/7   
3869                               Voirie         3   Accès réservé     24/7   
3870                               Voirie         3   Accès réservé     24/7   

     station_deux_roues     id_pdc_itinerance    date_maj  \
0                 false       FRSIGESIGE34211  2023-06-18   
1                 false       FRSIGESIGE34212  2023-06-18   
2                 false       FRSIGESIGE34412  2023-06-18   
3                 false       FRSIGESIGE34411  2023-06-18   
4                 false       FRS86ESOREB1511  2023-06-18   
...                 ...                   ...         ...   

- exemple : une station avec plusieurs adresses -> erreur de saisie ?

In [26]:
itinerance_5.loc[itinerance_5.coordonneesXY	 == '[2.87930851314442, 48.94679007929618]', relations]

index                        contact_operateur   nom_enseigne  \
113   9367  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
114   9368  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
115   9369  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
116   9370  supervision-ev.france@totalenergies.com  RIESTER MEAUX   

                             coordonneesXY  \
113  [2.87930851314442, 48.94679007929618]   
114  [2.87930851314442, 48.94679007929618]   
115  [2.87930851314442, 48.94679007929618]   
116  [2.87930851314442, 48.94679007929618]   

                           adresse_station id_station_itinerance  \
113  52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
114  52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
115  51 Rue François De Tessan 77100 Meaux           FRTCBP00702   
116  50 Rue François De Tessan 77100 Meaux           FRTCBP00702   

       nom_station          implantation_station  nbre_pdc condition_acces  \
113  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
114  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
115  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
116  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   

                                              horaires station_deux_roues  \
113  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
114  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
115  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
116  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   

    id_pdc_itinerance    date_maj                     last_modified  
113      FRTCBE003959  2023-01-27  2023-06-17T07:58:30.050000+00:00  
114      FRTCBE003958  2023-01-27  2023-06-17T07:58:30.050000+00:00  
115      FRTCBE003955  2023-01-27  2023-06-17T07:58:30.050000+00:00  
116      FRTCBE003954  2023-01-27  2023-06-17T07:58:30.050000+00:00

### Cohérence station - coordonnées
- 477 pdc sont associés à des stations avec plusieurs coordonnées

In [27]:
itinerance_4.loc[~itinerance_4['coordonneesXY - id_station_itinerance'], relations][200:215]

index  contact_operateur nom_enseigne              coordonneesXY  \
2199  75176  web@freshmile.com    Freshmile         [6.70933, 49.1065]   
2200  75185  web@freshmile.com    Freshmile         [6.70933, 49.1065]   
2201  75444  web@freshmile.com    Freshmile         [6.70929, 49.1065]   
2202  75699  web@freshmile.com    Freshmile  [7.49739994, 47.66957673]   
2203  75735  web@freshmile.com    Freshmile  [7.49751204, 47.66949847]   
2204  75840  web@freshmile.com    Freshmile      [1.523894, 44.147449]   
2205  75842  web@freshmile.com    Freshmile     [-0.672947, 44.859769]   
2207  75861  web@freshmile.com    Freshmile        [-1.15518, 44.6395]   
2210  75866  web@freshmile.com    Freshmile       [4.636316, 44.26687]   
2211  75869  web@freshmile.com    Freshmile        [0.66337, 43.85402]   
2212  75873  web@freshmile.com    Freshmile      [1.391725, 43.566937]   
2213  75882  web@freshmile.com    Freshmile      [6.082335, 43.243236]   
2229  75984  web@freshmile.com    Freshmile      [1.360146, 43.534457]   
2230  75987  web@freshmile.com    Freshmile      [4.836964, 43.888027]   
2231  76007  web@freshmile.com    Freshmile      [6.177948, 44.069971]   

                                adresse_station id_station_itinerance  \
2199          Place du Marché 57500 Saint-Avold          FRFR1PYKTKBS   
2200          Place du Marché 57500 Saint-Avold          FRFR1PYKTKBS   
2201          Place du Marché 57500 Saint-Avold          FRFR1PYKTKBS   
2202             Allée Eugène Moser 68680 Kembs      FRFR1PKCFNXIKPD8   
2203             Allée Eugène Moser 68680 Kembs      FRFR1PKCFNXIKPD8   
2204               90 avenue d'Embrun 05000 Gap      FRFR1PEHNKRRPQNR   
2205  87 Chemin de la Pépinière 13600 La Ciotat      FRFR1PD0AIJIWGHD   
2207  87 Chemin de la Pépinière 13600 La Ciotat      FRFR1PD0AIJIWGHD   
2210               90 avenue d'Embrun 05000 Gap      FRFR1PEHNKRRPQNR   
2211               90 avenue d'Embrun 05000 Gap      FRFR1PEHNKRRPQNR   
2212               90 avenue d'Embrun 05000 Gap      FRFR1PEHNKRRPQNR   
2213  87 Chemin de la Pépinière 13600 La Ciotat      FRFR1PD0AIJIWGHD   
2229     104 Avenue Docteur Jean Bru 47000 Agen      FRFR1PJWITZF0EBJ   
2230     104 Avenue Docteur Jean Bru 47000 Agen      FRFR1PJWITZF0EBJ   
2231  87 Chemin de la Pépinière 13600 La Ciotat      FRFR1PD0AIJIWGHD   

                                      nom_station implantation_station  \
2199                 Saint-Avold, place du Marché               Voirie   
2200                 Saint-Avold, place du Marché               Voirie   
2201                 Saint-Avold, place du Marché               Voirie   
2202                       Kembs, Rue De L'Europe               Voirie   
2203                       Kembs, Rue De L'Europe               Voirie   
2204               Gap, 90 avenue d'Embrun - 3273               Voirie   
2205  La Ciotat, 87 Chemin de la Pépinière - 3553               Voirie   
2207  La Ciotat, 87 Chemin de la Pépinière - 3553               Voirie   
2210               Gap, 90 avenue d'Embrun - 3273               Voirie   
2211               Gap, 90 avenue d'Embrun - 3273               Voirie   
2212               Gap, 90 avenue d'Embrun - 3273               Voirie   
2213  La Ciotat, 87 Chemin de la Pépinière - 3553               Voirie   
2229     Agen, 104 Avenue Docteur Jean Bru - 2840               Voirie   
2230     Agen, 104 Avenue Docteur Jean Bru - 2840               Voirie   
2231  La Ciotat, 87 Chemin de la Pépinière - 3553               Voirie   

      nbre_pdc condition_acces horaires station_deux_roues id_pdc_itinerance  \
2199         3     Accès libre     24/7              false       FRFR1ECJQR1   
2200         3     Accès libre     24/7              false       FRFR1ECJQR2   
2201         3     Accès libre     24/7              false       FRFR1EGYFJ1   
2202         2     Accès libre     24/7              false       FRFR1EWVHE1   
2203         2     Accès libre     24/7              false       FR

- exemple : une station avec plusieurs coordonnées -> incompréhension distinction station / pdc

In [28]:
itinerance_5.loc[itinerance_5.id_station_itinerance == 'FR073PCAMAIEUFR', relations]

index         contact_operateur    nom_enseigne         coordonneesXY  \
45   5377  acelec@acelec-france.com  Camaïeu France  [3.207187,50.684964]   
46   5379  acelec@acelec-france.com  Camaïeu France  [3.207433,50.684876]   
47   5380  acelec@acelec-france.com  Camaïeu France  [3.207433,50.684876]   
48   5381  acelec@acelec-france.com  Camaïeu France   [3.20737,50.684898]   
49   5382  acelec@acelec-france.com  Camaïeu France   [3.20737,50.684898]   
50   5383  acelec@acelec-france.com  Camaïeu France  [3.207306,50.684918]   
51   5384  acelec@acelec-france.com  Camaïeu France  [3.207306,50.684918]   
52   5385  acelec@acelec-france.com  Camaïeu France   [3.207246,50.68494]   
53   5386  acelec@acelec-france.com  Camaïeu France   [3.207246,50.68494]   
55   5388  acelec@acelec-france.com  Camaïeu France  [3.207187,50.684964]   
56   5389  acelec@acelec-france.com  Camaïeu France  [3.207124,50.684984]   
57   5390  acelec@acelec-france.com  Camaïeu France  [3.207047,50.685013]   
58   5391  acelec@acelec-france.com  Camaïeu France  [3.207047,50.685013]   
59   5392  acelec@acelec-france.com  Camaïeu France  [3.206962,50.685049]   
60   5393  acelec@acelec-france.com  Camaïeu France  [3.206962,50.685049]   
62   5395  acelec@acelec-france.com  Camaïeu France  [3.207124,50.684984]   

                       adresse_station id_station_itinerance     nom_station  \
45  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
46  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
47  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
48  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
49  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
50  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
51  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
52  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
53  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
55  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
56  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
57  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
58  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
59  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
60  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   
62  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR  CAMAÏEU FRANCE   

            implantation_station  nbre_pdc condition_acces           horaires  \
45  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
46  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
47  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
48  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
49  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
50  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
51  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
52  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
53  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
55  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
56  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
57  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
58  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
59  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
60  Parking privé à usage public         1     Accès libre  Mo-Su 08:00-08:00   
62  Parking privé à usage public     

- exemple : une station avec plusieurs coordonnées -> regroupement de stations ?

In [29]:
itinerance_5.loc[itinerance_5.id_station_itinerance == 'FR55CPBP514', relations]

index contact_operateur          nom_enseigne          coordonneesXY  \
3490  84105  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242799, 49.120583]   
3561  84197  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242799, 49.120583]   
3562  84198  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242799, 49.120583]   
3588  84293  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242644, 49.120556]   
3611  84319  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242644, 49.120556]   
3612  84320  contact@e55c.com  ELECTRIC 55 CHARGING  [4.242644, 49.120556]   

                            adresse_station id_station_itinerance  \
3490   BP CHAMPAGNE SUD - LES PETITES-LOGES           FR55CPBP514   
3561   BP CHAMPAGNE SUD - LES PETITES-LOGES           FR55CPBP514   
3562   BP CHAMPAGNE SUD - LES PETITES-LOGES           FR55CPBP514   
3588  BP CHAMPAGNE NORD - LES PETITES-LOGES           FR55CPBP514   
3611  BP CHAMPAGNE NORD - LES PETITES-LOGES           FR55CPBP514   
3612  BP CHAMPAGNE NORD - LES PETITES-LOGES           FR55CPBP514   

                                        nom_station  \
3490   BP CHAMPAGNE SUD - LES PETITES-LOGES - 69800   
3561   BP CHAMPAGNE SUD - LES PETITES-LOGES - 69800   
3562   BP CHAMPAGNE SUD - LES PETITES-LOGES - 69800   
3588  BP CHAMPAGNE NORD - LES PETITES-LOGES - 69800   
3611  BP CHAMPAGNE NORD - LES PETITES-LOGES - 69800   
3612  BP CHAMPAGNE NORD - LES PETITES-LOGES - 69800   

                     implantation_station  nbre_pdc condition_acces horaires  \
3490  Station dédiée à la recharge rapide         3   Accès réservé     24/7   
3561  Station dédiée à la recharge rapide         3   Accès réservé     24/7   
3562  Station dédiée à la recharge rapide         3   Accès réservé     24/7   
3588  Station dédiée à la recharge rapide         3   Accès réservé     24/7   
3611  Station dédiée à la recharge rapide         3   Accès réservé     24/7   
3612  Station dédiée à la recharge rapide         3   Accès réservé     24/7   

     station_deux_roues           id_pdc_itinerance    date_maj  \
3490              FALSE  FR55CE51400GIRVATMX02AA8A2  2022-11-03   
3561              FALSE  FR55CE51400GIRVATMX02AA8A1  2022-11-03   
3562              FALSE  FR55CE51400GIRVATMX02AA8A3  2022-11-03   
3588              FALSE  FR55CE51400GIRVATMX02A6A61  2022-11-03   
3611              FALSE  FR55CE51400GIRVATMX02A6B62  2022-11-03   
3612              FALSE  FR55CE51400GIRVATMX02A6B63  2022-11-03   

                         last_modified  
3490  2023-06-17T07:51:01.116000+00:00  
3561  2023-06-17T07:51:01.116000+00:00  
3562  2023-06-17T07:51:01.116000+00:00  
3588  2023-06-17T07:51:01.116000+00:00  
3611  2023-06-17T07:51:01.116000+00:00  
3612  2023-06-17T07:51:01.116000+00:00

- exemple : quatre stations avec une localisation identique mais deux adresses différentes + id identique entre station et pdc -> erreur d'adresse et erreur d'id_station ?

In [30]:
itinerance_5.loc[itinerance_5.coordonneesXY	 == '[-0.366184, 43.332508]', relations]

index contact_operateur       nom_enseigne           coordonneesXY  \
37   4472    sav@izivia.com  MOBIVE - SDEPA 64  [-0.366184, 43.332508]   
38   4473    sav@izivia.com  MOBIVE - SDEPA 64  [-0.366184, 43.332508]   
39   4474    sav@izivia.com  MOBIVE - SDEPA 64  [-0.366184, 43.332508]   
40   4475    sav@izivia.com  MOBIVE - SDEPA 64  [-0.366184, 43.332508]   

     adresse_station id_station_itinerance             nom_station  \
37  PARKING ZÉNITH 2        FRS64PMB649812  PAU - PARKING ZENITH 2   
38  PARKING ZÉNITH 2        FRS64PMB649811  PAU - PARKING ZENITH 2   
39  PARKING ZÉNITH 1        FRS64PMB649712  PAU - PARKING ZENITH 1   
40  PARKING ZÉNITH 1        FRS64PMB649711  PAU - PARKING ZENITH 1   

   implantation_station  nbre_pdc condition_acces horaires station_deux_roues  \
37               Voirie         2     Accès libre     24/7              false   
38               Voirie         2     Accès libre     24/7              false   
39               Voirie         2     Accès libre     24/7              false   
40               Voirie         2     Accès libre     24/7              false   

   id_pdc_itinerance    date_maj                     last_modified  
37    FRS64EMB649812  2023-06-18  2023-06-18T02:54:38.787000+00:00  
38    FRS64EMB649811  2023-06-18  2023-06-18T02:54:38.787000+00:00  
39    FRS64EMB649712  2023-06-18  2023-06-18T02:54:38.787000+00:00  
40    FRS64EMB649711  2023-06-18  2023-06-18T02:54:38.787000+00:00

### Incohérence 'nbre_pdc'
- 25% des pdc ont un champ 'nbre_pdc' mal documenté

- exemple : station avec un pdc 'de regroupement' qui évite de documenter plusieurs lignes

In [31]:
itinerance_init.loc[itinerance_init.id_station_itinerance == 'FRLUMEACACIAS11', 
               ['id_pdc_itinerance', 'id_station_itinerance', 'coordonneesXY', 'nbre_pdc']]

id_pdc_itinerance id_station_itinerance           coordonneesXY  \
64380   FRLUMEACACIAS11       FRLUMEACACIAS11  [3.1485441,50.7582304]   

       nbre_pdc  
64380         4

- exemple : station avec plusieurs pdc mais l'id_station est identique à l'id_pdc

In [32]:
itinerance_init.loc[itinerance_init.coordonneesXY	 == '[2.451322, 45.66523]', 
                   ['id_pdc_itinerance', 'id_station_itinerance', 'coordonneesXY', 'nbre_pdc']]

id_pdc_itinerance id_station_itinerance         coordonneesXY  nbre_pdc
5239     FRSDGEMBDG512         FRSDGPMBDG512  [2.451322, 45.66523]         2
5240     FRSDGEMBDG511         FRSDGPMBDG511  [2.451322, 45.66523]         2

- exemple : station avec plusieurs pdc mais le champs a toujours une valeur de 1

In [33]:
itinerance_init.loc[itinerance_init.coordonneesXY	 == '[-1.7548354193520386, 48.125067030488154]', 
                   ['id_pdc_itinerance', 'id_station_itinerance', 'coordonneesXY', 'nbre_pdc']]

id_pdc_itinerance id_station_itinerance  \
82400    FRIENE35353A51          FRIENS35353A   
82401    FRIENE35353A41          FRIENS35353A   
82402    FRIENE35353A31          FRIENS35353A   
82403    FRIENE35353A21          FRIENS35353A   
82404    FRIENE35353A11          FRIENS35353A   

                                   coordonneesXY  nbre_pdc  
82400  [-1.7548354193520386, 48.125067030488154]         1  
82401  [-1.7548354193520386, 48.125067030488154]         1  
82402  [-1.7548354193520386, 48.125067030488154]         1  
82403  [-1.7548354193520386, 48.125067030488154]         1  
82404  [-1.7548354193520386, 48.125067030488154]         1

### cohérence station - enseigne
- exemple de station avec plusieurs noms d'enseigne

In [34]:
itinerance_init.loc[itinerance_init.id_station_itinerance == 'FRS27PBARREOUCHEMOMORT', relations]

index            contact_operateur nom_enseigne         coordonneesXY  \
65099  65411  pascal.lhermitte@siege27.fr     SIEGE266  [0.66299, 48.946159]   
65100  65412  pascal.lhermitte@siege27.fr     SIEGE265  [0.66299, 48.946159]   

                                      adresse_station   id_station_itinerance  \
65099   Place du monument aux Morts La Barre en Ouche  FRS27PBARREOUCHEMOMORT   
65100   Place du monument aux Morts La Barre en Ouche  FRS27PBARREOUCHEMOMORT   

      nom_station implantation_station  nbre_pdc condition_acces horaires  \
65099      900120               Voirie         2     Accès libre     24/7   
65100      900120               Voirie         2     Accès libre     24/7   

      station_deux_roues        id_pdc_itinerance    date_maj  \
65099              false  FRS27EBARREOUCHEMOMORTG  2021-06-15   
65100              false  FRS27EBARREOUCHEMOMORTD  2021-06-15   

                          last_modified  
65099  2023-06-17T07:55:06.631000+00:00  
65100  2023-06-17T07:55:06.631000+00:00

### cohérence station - horaires
- exemple de station avec plusieurs types d'horaire

In [35]:
itinerance_init.loc[itinerance_init.id_station_itinerance == 'FRS23D2302001', relations]

index  contact_operateur               nom_enseigne  \
64653  64958  contact@mobive.fr  Réseau de recharge Creuse   
64654  64959  contact@mobive.fr  Réseau de recharge Creuse   

                                coordonneesXY adresse_station  \
64653  [2.297630254367735, 45.98220366857867]  Place du Treix   
64654  [2.297630254367735, 45.98220366857867]  Place du Treix   

      id_station_itinerance           nom_station implantation_station  \
64653         FRS23D2302001  BELLEGARDE EN MARCHE       Parking public   
64654         FRS23D2302001  BELLEGARDE EN MARCHE       Parking public   

       nbre_pdc condition_acces           horaires station_deux_roues  \
64653         2     Accès libre  Mo-Su 00:00-00:01              False   
64654         2     Accès libre  Mo-Su 00:00-00:02              False   

      id_pdc_itinerance    date_maj                     last_modified  
64653   FRS23P230200101  2023-03-09  2023-06-17T07:55:10.242000+00:00  
64654   FRS23P230200102  2023-03-09  2023-06-17T07:55:10.242000+00:00

### cohérence station - deux-roues
- exemple de station avec plusieurs indications d'accessibilité deux-roues

In [36]:
itinerance_init.loc[itinerance_init.id_station_itinerance == 'FRV75PPX1902', relations]

index                        contact_operateur nom_enseigne  \
10044  10154  supervision-ev.france@totalenergies.com       Belib'   
10045  10155  supervision-ev.france@totalenergies.com       Belib'   
10066  10176  supervision-ev.france@totalenergies.com       Belib'   
10067  10177  supervision-ev.france@totalenergies.com       Belib'   
10068  10178  supervision-ev.france@totalenergies.com       Belib'   
10069  10179  supervision-ev.france@totalenergies.com       Belib'   
10070  10180  supervision-ev.france@totalenergies.com       Belib'   

                coordonneesXY                    adresse_station  \
10044  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
10045  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
10066  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
10067  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
10068  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
10069  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    
10070  [2.3730799, 48.888195]  61 Avenue de Flandre 75019 Paris    

      id_station_itinerance                   nom_station  \
10044          FRV75PPX1902  Paris | Avenue de Flandre 61   
10045          FRV75PPX1902  Paris | Avenue de Flandre 61   
10066          FRV75PPX1902  Paris | Avenue de Flandre 61   
10067          FRV75PPX1902  Paris | Avenue de Flandre 61   
10068          FRV75PPX1902  Paris | Avenue de Flandre 61   
10069          FRV75PPX1902  Paris | Avenue de Flandre 61   
10070          FRV75PPX1902  Paris | Avenue de Flandre 61   

      implantation_station  nbre_pdc condition_acces horaires  \
10044               Voirie         7     Accès libre     24/7   
10045               Voirie         7     Accès libre     24/7   
10066               Voirie         7     Accès libre     24/7   
10067               Voirie         7     Accès libre     24/7   
10068               Voirie         7     Accès libre     24/7   
10069               Voirie         7     Accès libre     24/7   
10070               Voirie         7     Accès libre     24/7   

      station_deux_roues id_pdc_itinerance    date_maj  \
10044              False     FRV75EPX19027  2023-06-02   
10045              False     FRV75EPX19026  2023-06-02   
10066              False     FRV75EPX19021  2023-06-02   
10067               True     FRV75EPX19022  2023-06-02   
10068               True     FRV75EPX19023  2023-06-02   
10069              False     FRV75EPX19024  2023-06-02   
10070              False     FRV75EPX19025  2023-06-02   

                          last_modified  
10044  2023-06-17T07:58:29.780000+00:00  
10045  2023-06-17T07:58:29.780000+00:00  
10066  2023-06-17T07:58:29.780000+00:00  
10067  2023-06-17T07:58:29.780000+00:00  
10068  2023-06-17T07:58:29.780000+00:00  
10069  2023-06-17T07:58:29.780000+00:00  
10070  2023-06-17T07:58:29.780000+00:00

### cohérence coordonnées - adresse
- exemple de plusieurs stations avec des coordonnées identiques mais des adresses différentes

In [37]:
itinerance_5.loc[itinerance_5.coordonneesXY == '[2.460441, 50.78763]', relations]

index              contact_operateur          nom_enseigne  \
1396  51613         support@alizecharge.fr  pass pass électrique   
1397  51617         support@alizecharge.fr  pass pass électrique   
1427  52778         support@alizecharge.fr  pass pass électrique   
1431  52983         support@alizecharge.fr  pass pass électrique   
1584  64885  support@passpasselectrique.fr  Pass Pass Electrique   

             coordonneesXY                       adresse_station  \
1396  [2.460441, 50.78763]            Pole Gare 59670 BAVINCHOVE   
1397  [2.460441, 50.78763]            Pole Gare 59670 BAVINCHOVE   
1427  [2.460441, 50.78763]            Pole Gare 59670 BAVINCHOVE   
1431  [2.460441, 50.78763]            Pole Gare 59670 BAVINCHOVE   
1584  [2.460441, 50.78763]  1bis Rue De La Gare 59670 Bavinchove   

     id_station_itinerance             nom_station implantation_station  \
1396        FRH14P59054001  BAVINCHOVE - Pole Gare               Voirie   
1397        FRH14P59054001  BAVINCHOVE - Pole Gare               Voirie   
1427        FRH14P59054001  BAVINCHOVE - Pole Gare               Voirie   
1431        FRH14P59054001  BAVINCHOVE - Pole Gare               Voirie   
1584        FRH14E59054001       Bavinchove - Gare       Parking public   

      nbre_pdc condition_acces           horaires station_deux_roues  \
1396         4     Accès libre  Mo-Su 00:00-23:59              FALSE   
1397         4     Accès libre  Mo-Su 00:00-23:59              FALSE   
1427         4     Accès libre  Mo-Su 00:00-23:59              FALSE   
1431         4     Accès libre  Mo-Su 00:00-23:59              FALSE   
1584         2     Accès libre               24/7              false   

     id_pdc_itinerance    date_maj                     last_modified  
1396   FRH14E590540013  2023-03-17  2023-06-17T07:55:57.156000+00:00  
1397   FRH14E590540012  2023-03-17  2023-06-17T07:55:57.156000+00:00  
1427   FRH14E590540014  2023-03-17  2023-06-17T07:55:57.156000+00:00  
1431   FRH14E590540011  2023-03-17  2023-06-17T07:55:57.156000+00:00  
1584    FRH14E59054001  2023-01-27  2023-06-17T07:55:13.011000+00:00